manuals/odrawverify.1
.Dd June 27, 2026
.Dt ODRAWVERIFY 1
.Os
.Sh NAME
.Nm odrawverify
.Nd verifies media data stored in an optical disc (split) RAW image file
.Sh SYNOPSIS
.Nm odrawverify
.Op Fl d Ar digest_type
.Op Fl l Ar log_file
.Op Fl p Ar buffer_size
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm odrawverify
is a utility to verify media data stored in an optical disc (split) RAW image \
file
.Pp
.Nm odrawverify
is part of the
.Nm libodraw
package.
.Nm libodraw
is a library to access optical disc (split) RAW formats
.Pp
.Ar source
is the source image.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl d Ar digest_type
calculate additional digest (hash) types besides md5, options: sha1, sha256
.It Fl h
shows this help
.It Fl l Ar log_file
logs verification errors and the digest (hash) to a file
.It Fl p Ar buffer_size
specify the process buffer size (default is the 32768)
.It Fl q
quiet shows minimal status information
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# odrawverify cd.cue
odrawverify 20110801
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr odrawinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libodraw/issues
.Sh COPYRIGHT
Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO:
* fix verbose/debug output issue in IO handle regarding sector mode
* add Python bindings
* add thread safely

* find a solution for
  - redefinition of _FILE_OFFSET_BITS for Solaris
  - redefinition of _MIN and _MAX for BorlandC
* library
  - add read error support
  - add run-out support (partial)
  - track data cache
  - cache current sector in IO handle
  - check MSF in track sector data (requires current sector)
  - emulate checksum and ECC data for 2352 byte sectors
* scanner
  - put syntax error in error object
  - use start conditions %s for REM (has impact on parser)
  - different types of ISRC definitions
* parser
  - make sure sector offsets are relative from the start of the media
  - handle success file commands without a track command
  - refactor code into separate functions?
* session
  - handle pregap and postgap
  - add default 1 session if no session data is available ?
* cue file
  - add syntax restrictions
  - add support for single digit values
  - handle reset of track MSF to 00:00:00
* add support for .toc file
* libodraw_track_value_copy_track_data_to_buffer
  - add read error support
  - emulate sector error correction data

Tools:
* odrawinfo
  - add flag to skip open of data files ?
  - add byte size string output (see smdevinfo)
* odrawmount
  - create?
  - mount different sessions

Deployment
* check man pages
* check functions included in configure.ac

Tests:

20130831
* see `git log' for more recent change log
* updated dependencies
* removed examples
* added macosx files
* worked on tests
* added support for codepage 949, 950 and 1255
* updated msvscpp files
* code clean up

20130721
* updated dependencies

20130329
* small changes

20130327
* 2013 update
* updated dependencies
* changes for project move

20120923
* updated dependencies
* code clean up

20120701
* changes due to updates
* updated msvscpp files

20120630
* updated dependencies

20120212
* small changes

20120209
* updated libnotify, libbfio
* added dpkg files

20120102
* 2012 update

20111231
* updated libcstring, libuna

20111224
* updated libcstring, libsystem, libuna
* updated array type
* code clean up

20111102
* updated m4 files
* updated libcstring, libsystem, libuna, libbfio

20111030
* updated configure.as and m4 files
* removed libfvalue

20111008
* updated configure.as and m4 files
* updated README files
* updated pc and spec file
* updated common
* updated libuna, libbfio, libsystem

20110918
* updated configure

20110914
* updated configure

20110831
* updated configure, libuna

20110806
* updated configure.ac and acinclude.m4
* updated libfvalue

20110726
* worked on support for handle reset of track MSF to 00:00:00 after new FILE
* updated msvscpp files

20110725
* updated common, libsystem, libfvalue,
* added support for blue book cue, multiple audio tracks per session
* odrawinfo added option to ignore data files (-i)
* fixed print of last sector when number of sectors is 0
* worked on support for handle reset of track MSF to 00:00:00 after new FILE

20110625
* changes for: warning: AC_LANG_CONFTEST: no AC_LANG_SOURCE call detected in body
* code clean up
* updated common, libuna, libbfio, libfvalue

20110510
* small correction in verbose output

20110508
* updated configure.ac
* updated include/codepage.h
* updated libcstring, libsystem
* added INTLLIBS to odrawtools Makefile.am
* worked on sector header sanity checks

20110312
* updated libcstring, libuna, libbfio, libsystem
* bug fix in session size correction

20110211
* synced with secondary branch
* fixed multiple issues in CUE parser

20110209
* added support for audio tracks

20110203
* updated libsystem

20110114
* updated libsystem, libuna, libbfio

20110114
* updated common, libsystem, libuna, libbfio

20110111
* updated libcstring, libuna, libbfio, libfvalue
* updated man pages
* updated msvscpp files
* fixes for mingw compilation

20110109
* worked on mode 2 XA form-1 support
* changed LIBODRAW_TRACK_TYPE_MODE2_2342 into LIBODRAW_TRACK_TYPE_MODE2_2324
* updated configure.ac
* updated common, libcstring, libsystem, libuna, libbfio
* debug output of scanner and parser now uses libnotify

20110106
* small fixes in header files
* data_file_set_filename overwrites auto-detection behavior of open_data_files
* refactore copy track data to IO handle

20110105
* worked on seek and read tests
* worked on non-sector size read buffer
* read buffer added read beyond media size support
* fixed leak in cue scanner of yy_buffer_stack, by not calling cue scanner destroy

20110104
* updated libsystem, libuna, libbfio, libfvalue
* 2011 update
* worked on seek offset
* added initial run-out support
* added tests directory
* added example directory
* worked on seek and read tests

20101221
* worked on read buffer

20101220
* worked on media offset in lead-outs, sessions, tracks, data_file
* removed list type and sector list
* worked on read buffer

20101219
* updated array type
* updated libsystem, libbfio, libfvalue
* code clean up
* worked on data file support
* worked on determining media size and sizes missing from CUE

20101218
* fix for mingw compilation

20101217
* renamed libodtoc to libodraw
* updated configure.ac, added gettext support
* updated array type, list type, sector list
* updated common, liberror, libnotify, libsystem, libuna, libbfio, libfvalue
* added unused
* renamed file to handle
* added %no-lines to yacc file for Builder5 support

20101119
* updated configure.ac
* updated libcsting, liberror
* added local library support

20101117
* worked on file support

20101116
* added file support
* changed track support to entail file and type

20101115
* updated libbfio
* worked on initial version with CUE support
  - checked msvscpp build
  - overwritten lex/yacc memory management
  - turned off lex file options
* added list type and offset list
* added runtime information for sessions and tracks

20101114
* worked on initial version with CUE support
  - created lex/yacc parser skeleton

20101113

odrawtools/odrawverify.c
/*
 * Verifies the integrity of optical disc (split) RAW file(s)
 *
 * Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#include "byte_size_string.h"
#include "digest_hash.h"
#include "log_handle.h"
#include "odrawtools_getopt.h"
#include "odrawtools_libcerror.h"
#include "odrawtools_libclocale.h"
#include "odrawtools_libcnotify.h"
#include "odrawtools_libodraw.h"
#include "odrawtools_output.h"
#include "odrawtools_signal.h"
#include "odrawtools_unused.h"
#include "verification_handle.h"

verification_handle_t *odrawverify_verification_handle = NULL;
int odrawverify_abort                                  = 0;

/* Signal handler for odrawverify
 */
void odrawverify_signal_handler(
      odrawtools_signal_t signal ODRAWTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function   = "odrawverify_signal_handler";

	ODRAWTOOLS_UNREFERENCED_PARAMETER( signal )

	odrawverify_abort = 1;

	if( odrawverify_verification_handle != NULL )
	{
		if( verification_handle_signal_abort(
		     odrawverify_verification_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal verification handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use odrawverify to verify data stored in the optical disc (split) RAW image file format.";

	odrawtools_option_t options[ ] = {
		{ 'd', "digest_type", "calculate additional digest (hash) types besides md5, options: sha1, sha256" },
		{ 'h', NULL, "shows this help" },
		{ 'l', "log_file", "logs verification errors and the digest (hash) to a file" },
		{ 'p', "buffer_size", "specify the process buffer size (default is the 32768)" },
		{ 'q', NULL, "quiet shows minimal status information" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source image" },
	};
	system_character_t options_string[ 32 ];

	libcerror_error_t *error                           = NULL;
	log_handle_t *log_handle                           = NULL;
	system_character_t *log_filename                   = NULL;
	system_character_t *option_additional_digest_types = NULL;
	system_character_t *option_process_buffer_size     = NULL;
	system_character_t *source                         = NULL;
	system_integer_t option                            = 0;
	char *program                                      = "odrawverify";
	uint8_t calculate_md5                              = 1;
	int number_of_options                              = (int) ( sizeof( options ) / sizeof( odrawtools_option_t ) );
	uint8_t print_status_information                   = 1;
	uint8_t verbose                                    = 0;
	int result                                         = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
             "odrawtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( odrawtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	odrawtools_output_version_fprint(
	 stdout,
	 program );

	if( odrawtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = odrawtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				odrawtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				goto on_error;

			case (system_integer_t) 'd':
				option_additional_digest_types = optarg;

				break;

			case (system_integer_t) 'h':
				odrawtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'l':
				log_filename = optarg;

				break;

			case (system_integer_t) 'p':
				option_process_buffer_size = optarg;

				break;

			case (system_integer_t) 'q':
				print_status_information = 0;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				odrawtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		odrawtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );

#if !defined( HAVE_LOCAL_LIBODRAW )
	libodraw_notify_set_verbose(
	 verbose );
	libodraw_notify_set_stream(
	 stderr,
	 NULL );
#endif

	if( verification_handle_initialize(
	     &odrawverify_verification_handle,
	     calculate_md5,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to create verification handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( odrawverify_verification_handle != NULL );
#endif
	if( option_process_buffer_size != NULL )
	{
		result = verification_handle_set_process_buffer_size(
			  odrawverify_verification_handle,
			  option_process_buffer_size,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set process buffer size.\n" );

			goto on_error;
		}
		else if( ( result == 0 )
		      || ( odrawverify_verification_handle->process_buffer_size > (size_t) SSIZE_MAX ) )
		{
			odrawverify_verification_handle->process_buffer_size = 0;

			fprintf(
			 stderr,
			 "Unsupported process buffer size defaulting to: 32768.\n" );
		}
	}
	if( option_additional_digest_types != NULL )
	{
		result = verification_handle_set_additional_digest_types(
			  odrawverify_verification_handle,
			  option_additional_digest_types,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set additional digest types.\n" );

			goto on_error;
		}
	}
	if( odrawtools_signal_attach(
	     odrawverify_signal_handler,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to attach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( verification_handle_open_input(
	     odrawverify_verification_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open file: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( log_filename != NULL )
	{
		if( log_handle_initialize(
		     &log_handle,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to create log handle.\n" );

			goto on_error;
		}
		if( log_handle_open(
		     log_handle,
		     log_filename,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to open log file: %" PRIs_SYSTEM ".\n",
			 log_filename );

			goto on_error;
		}
	}
	result = verification_handle_verify_input(
		  odrawverify_verification_handle,
		  print_status_information,
		  log_handle,
		  &error );

	if( result != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to verify input.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( log_handle != NULL )
	{
		if( log_handle_close(
		     log_handle,
		     &error ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to close log handle.\n" );

			goto on_error;
		}
		if( log_handle_free(
		     &log_handle,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to free log handle.\n" );

			goto on_error;
		}
	}
	if( odrawtools_signal_detach(
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to detach signal handler.\n" );

		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( verification_handle_close(
	     odrawverify_verification_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close verification handle.\n" );

		goto on_error;

	}
	if( verification_handle_free(
	     &odrawverify_verification_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free verification handle.\n" );

		goto on_error;
	}
	if( odrawverify_abort != 0 )
	{
		fprintf(
		 stdout,
		 "%" PRIs_SYSTEM ": ABORTED\n",
		 program );

		return( EXIT_FAILURE );
	}
	if( result != 1 )
	{
		fprintf(
		 stdout,
		 "%" PRIs_SYSTEM ": FAILURE\n",
		 program );

		return( EXIT_FAILURE );
	}
	fprintf(
	 stdout,
	 "%" PRIs_SYSTEM ": SUCCESS\n",
	 program );

	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( log_handle != NULL )
	{
		log_handle_close(
		 log_handle,
		 NULL );
		log_handle_free(
		 &log_handle,
		 NULL );
	}
	if( odrawverify_verification_handle != NULL )
	{
		verification_handle_close(
		 odrawverify_verification_handle,
		 NULL );
		verification_handle_free(
		 &odrawverify_verification_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

